flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,270 @@
+
+        private final Map<ExecutionVertexID, ExecutionSlotSharingGroup>
+                executionSlotSharingGroupMap;
+
+        private final Map<JobVertexID, CoLocationGroup> coLocationGroupMap;
+
+        private final Map<CoLocationConstraint, ExecutionSlotSharingGroup> clcToESsgMap;
+
+        private TaskBalancedExecutionSlotSharingGroupBuilder(
+                final SchedulingTopology topology,
+                final Set<SlotSharingGroup> logicalSlotSharingGroups,
+                final Set<CoLocationGroup> coLocationGroups) {
+            this.topology = checkNotNull(topology);
+
+            this.coLocationGroupMap = new HashMap<>();
+            for (CoLocationGroup coLocationGroup : coLocationGroups) {
+                for (JobVertexID jobVertexId : coLocationGroup.getVertexIds()) {
+                    coLocationGroupMap.put(jobVertexId, coLocationGroup);
+                }
+            }
+
+            this.clcToESsgMap = new HashMap<>();
+            this.ssgToExecutionSSGs = new HashMap<>(logicalSlotSharingGroups.size());
+            this.slotRoundRobinIndex = new HashMap<>(logicalSlotSharingGroups.size());
+            this.slotSharingGroupMap = new HashMap<>();
+            this.executionSlotSharingGroupMap = new HashMap<>();
+
+            for (SlotSharingGroup slotSharingGroup : logicalSlotSharingGroups) {
+                for (JobVertexID jobVertexId : slotSharingGroup.getJobVertexIds()) {
+                    slotSharingGroupMap.put(jobVertexId, slotSharingGroup);
+                }
+            }
+        }
+
+        private Map<ExecutionVertexID, ExecutionSlotSharingGroup> build() {
+            final LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> allVertices =
+                    getExecutionVertices(topology);
+
+            computeSsgToESsgMap(allVertices);
+
+            // Loop on job vertices
+            for (Map.Entry<JobVertexID, List<SchedulingExecutionVertex>> entry :
+                    allVertices.entrySet()) {
+                SlotSharingGroup ssg = slotSharingGroupMap.get(entry.getKey());
+                List<ExecutionSlotSharingGroup> executionSSGs = ssgToExecutionSSGs.get(ssg);
+                int index = getSlotRoundRobinIndex(entry.getValue().size(), ssg);
+
+                for (SchedulingExecutionVertex sev : entry.getValue()) {
+                    // For no CoLocations or with CoLocations with max parallelism of the SSG.
+                    if (!coLocationGroupMap.containsKey(entry.getKey())
+                            || ssgToExecutionSSGs.get(ssg).size() == entry.getValue().size()) {
+                        index = addVertexToExecutionSlotSharingGroup(executionSSGs, index, sev);
+                        continue;
+                    }
+
+                    // For CoLocations with exited execution slot sharing group.
+                    final CoLocationConstraint clc = getCoLocationConstraint(entry.getKey(), sev);
```

Comment:
Why there should be a colocation group for this vertex? What about vertices whose parallelism less than the size of essg?